client/my-sites/theme/theme-style-variations/index.tsx

```diff
@@ -0,0 +1,99 @@
+import { Button } from '@automattic/components';
+import { PremiumBadge } from '@automattic/design-picker';
+import { useLayoutEffect, useRef, useState } from '@wordpress/element';
+import classNames from 'classnames';
+import { translate } from 'i18n-calypso';
+import AsyncLoad from 'calypso/components/async-load';
+import type { StyleVariation } from '@automattic/design-picker/src/types';
+import type { TranslateResult } from 'i18n-calypso';
+
+import './style.scss';
+
+interface ThemeStyleVariationsProps {
+	description: TranslateResult;
+	selectedVariation: StyleVariation;
+	variations: StyleVariation[];
+	onClick: ( variation: StyleVariation ) => void;
+}
+
+const ThemeStyleVariations = ( {
+	description,
+	selectedVariation,
+	variations,
+	onClick,
+}: ThemeStyleVariationsProps ) => {
+	const observerRef = useRef< HTMLDivElement | null >( null );
+	const [ collapsibleMaxHeight, setCollapsibleMaxHeight ] = useState< number | undefined >();
+	const [ isCollapsible, setIsCollapsible ] = useState< boolean >( false );
+	const [ isCollapsed, setIsCollapsed ] = useState< boolean >( false );
+
+	useLayoutEffect( () => {
+		if ( ! observerRef.current ) {
+			return;
+		}
+
+		const resizeObserver = new ResizeObserver( ( [ observerNode ] ) => {
+			const node = observerNode.target;
+			const nodeFirstChildRect = ( node.firstChild as HTMLElement )?.getBoundingClientRect();
+			const nodeLastChildRect = ( node.lastChild as HTMLElement )?.getBoundingClientRect();
+			if ( ! nodeFirstChildRect || ! nodeLastChildRect ) {
+				return;
+			}
+
+			// Detect flex wrap.
+			const currentIsCollapsible = nodeFirstChildRect.top !== nodeLastChildRect.top;
+			const shouldCollapse = isCollapsible !== currentIsCollapsible || isCollapsed;
+
+			setIsCollapsible( currentIsCollapsible );
+			setIsCollapsed( shouldCollapse );
+			setCollapsibleMaxHeight( shouldCollapse ? nodeFirstChildRect.height : node.scrollHeight );
+		} );
+
+		resizeObserver.observe( observerRef.current );
+		return () => {
+			resizeObserver.disconnect();
+		};
+	}, [ isCollapsible, isCollapsed ] );
```

Comment:
We can use `useRef` to get the latest value in the callback of the resize observer

Reply:
Could you explain more? Do you mean to use `useRef` instead of `useState` for `isCollapsible` and `isCollapsed`? Perhaps I'm missing something, but don't we need to update the state since we need to re-render the component (adding CSS classes, etc)? 🤔

Reply:
I mean we can create a new variable and its value is from `useRef` so that we can avoid the resize observer disconnecting and reconnecting whenever the state changes.

```js
const isCollapsibleRef = useRef();
const isCollapsedRef = useRef();

isCollapsibleRef.current = isCollapsible;
isCollapsedRef.current = isCollapsed;
```

What do you think?

Reply:
Hmmm, I managed to use this approach for `isCollapsible` but not for `isCollapsed`. I think it doesn't work for `isCollapsed` since the `ResizeObserver` callback isn't called when the state changes. Whereas it works for `isCollapsible` since it only changes state when resizing, so the `ResizeObserver` callback is called.

Am I missing anything?

Reply:
I think we should update the value of ref outside the `useEffect`

Reply:
@arthur791004 I've updated the code to what I _think_ is what you meant. It's not my usual coding approach, so let me know if it looks good to you!